### PR TITLE
LPS-130263 Balloon editor text toolbar's configuration

### PR DESCRIPTION
**Pull request by ambrinchaudhary**
Hey @julien, I think we should follow the pattern we have in other editors about how to pass the config to the BalloonToolbar through the `ConfigContributors`. 

Please have a look an tell me your impressions, note that this is still a WIP and I basically moved your js config to the java classes, but it doesn't mean they fit the toolbar buttons defined in the story ;-)

### CODE REVIEW

**Comment by ambrinchaudhary** [comment]
ci:test:sf

**Comment by julien** [comment]
@ambrinchaudhary, If I understand correctly, this gives us a way of configuring the editor's toolbars and plugins via Config Contributors so this **looks good to me**.

The only doubt I had was about the `extraPlugins`: because without the `floatingspace` and `balloontoolbar` plugins, our "BalloonEditor" will not work (at least not as a balloon editor), so I'm not sure if they should be in the config contributor or always in the React component, with a way of adding more via config contributor; but this is non blocking and something we can discuss.

And the last thing we might want to do before setting this as "ready to review" is to rebase with `master` when possible

**Comment by ambrinchaudhary** [comment]
> The only doubt I had was about the extraPlugins: because without the floatingspace and balloontoolbar plugins, our "BalloonEditor" will not work (at least not as a balloon editor), so I'm not sure if they should be in the config contributor or always in the React component, with a way of adding more via config contributor;

About this, we are doing the same in other editors. Required plugins are being added also via config contributors (for example [Creole editor](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/CKEditorCreoleConfigContributor.java#L76) does not work without the `creole` plugin), knowing that this contributor is always applied for our editor. In this particular case, every balloon editor will pass through this because of https://github.com/liferay-frontend/liferay-portal/pull/960/commits/b5664707c8a9d87126dfc81ecea89e46e9cfba41#diff-91a08522aea78564ab2061bea292615577dc05c3db94c579dd042dd8c5107544R30. And those plugins will be added for sure. So I still think we should only keep all our plugins at one single place, which are the config contributors. 

Sure, will rebase once ready.... but taking in mind that the original pull has been reverted [here](https://github.com/liferay-frontend/liferay-portal/pull/964), I will wait for that.

**Comment by julien** [comment]
@ambrinchaudhary what I meant is that if you use `BalloonEditor` via JS only you won't get the required plugins - ant that's what we want to be able to do, so I wouldn't compare it to the Creole editor.

**Comment by julien** [comment]
@ambrinchaudhary concerning the original pull request, it's still available [here](https://github.com/julien/liferay-portal/tree/LPS-127012) (at least part of it: we will need to take into account Brian's fix-up commits: see @wincent's [comment](https://github.com/brianchandotcom/liferay-portal/pull/100712#issue-613666005)) , so you can still fetch it - *and since you have been invited* to [my liferay-portal fork](https://github.com/julien/liferay-portal) when we started working on this you should be able to push too.

If you need any help,  don't hesitate to ask.

In anycase, I'll be working on fixing the missing (and breaking) code for [127012](https://issues.liferay.com/browse/LPS-127012) and will send a PR during the day.

**Comment by ambrinchaudhary** [comment]
> @ambrinchaudhary what I meant is that if you use BalloonEditor via JS only you won't get the required plugins - ant that's what we want to be able to do, so I wouldn't compare it to the Creole editor.

Good point @julien. You are rigth, if we use the `BalloonEditor` component diirectly via JS this `ConfigContributor` does not automatically return you the config, but you have to call for it manually. This is something that is already happening in Data Engine I think (at least I was told that), and in Lima they are doing the same with the ClassicEditor [here](https://github.com/liferay-lima/liferay-portal/pull/1736/commits/8903a36253dd5c07175b8d750be53704ad2e2a4c). 
So, when rendering the editor via JS, you need to manually get the config via 

```
EditorConfiguration editorConfiguration =
	EditorConfigurationFactoryUtil.getEditorConfiguration(
		TranslationPortletKeys.TRANSLATION, "translateEditor",
		"ckeditor", inputEditorTaglibAttributes, _themeDisplay,
		RequestBackedPortletURLFactoryUtil.create(_httpServletRequest));
```

So, we can have only keep one single point of configuration. Anyway, if you still that those plugins should be better placed in the js, we can move them. But to get the rest of the config (like the toolbar buttons...) access the `ConfigContributor` is needed.

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js [hunk 1]
```diff
@@ -20,74 +20,33 @@ import {Editor} from './Editor';
...
-			...config,
-			extraAllowedContent: '*',
-			extraPlugins: `${extraPlugins}${defaultExtraPlugins}`,
-		};
-	};
-
 	return (
 		<Editor
-			config={getConfig()}
+			config={config}
```
Comment (julien) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:25]: @ambrinchaudhary I'm still not sure about this approach so it would be nice to clarify if we want to be able to use this component without Java ... but maybe there's something I don't understand

For example: What if you want your React component to render `BalloonEditor`?
Does that mean you have to "glue" something in between to get the config from the config contributor?

Comment (ambrinchaudhary) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:25]: Yes, that is what I meant [here](https://github.com/liferay-frontend/liferay-portal/pull/960#issuecomment-818595271). You need to get the config via Java, and then pass it to your react component. Lima is already doing it, and I think DE too. This way, you still have one single point to configure your editor (the _*ConfigContributors.java_), and the editor could be rendered via taglib or via JS.

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/CKEditorBalloonEditorConfigContributor.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+			getToolbarTextJSONObject());
+
+		jsonObject.put("toolbars", toolbarsJSONArray);
+	}
+
+	protected JSONObject getToolbarImageJSONObject() {
+		return JSONUtil.put(
+			"buttons", "JustifyLeft,JustifyCenter,JustifyRight,Link,Unlink"
+		).put(
+			"priority", Boolean.TRUE
```
[the remark this answers is not in the document]
Comment (ambrinchaudhary) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/CKEditorBalloonEditorConfigContributor.java:130]: Yes, sure, but as in Java I don't have those values, for now this is only a flag and then in js we are changing to a number value. As for now we are only setting the HIGH priority, I choosed a bollean value, but this is something we can change if needed.

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js (hunk 1 above)
Comment (julien) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:25]: Thanks for making that clear.

Now I have another question, is this what **we** really want?

You said that Lima is doing it and DE are doing it, but that's doesn't answer my question: as a (**frontend**) infrastructure team, I think that we should take this decision based on what we think is going to be easier/better for other teams (and more specifically for the frontend developers working in other teams because they are the ones who are going to use this component) and not the other way around. At the end of the day, each team uses our infrastructure they way they want to use it (this is not a negative comment, it's just what I've observed and it makes sense since every team has their use cases), but I really think that the important point here is that the easier we make it the happier they will be (and so will we, because we have to maintain this).

So here's another take on my question: as a frontend developer, do you think depending on Java to get a the configuration for React component is ideal?

I personally think that would be a **bad idea**, because it makes things more complex.
Maybe you're used to reading Java code but I think most **frontend** developers aren't (and I'm including myself). 

Of course, if  the technical requirements specifies that the configuration must be written in Java (I'll have a look too) we can do it, but I just think that's just adding unnecessary complexity.

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/CKEditorBalloonEditorConfigContributor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (julien) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/CKEditorBalloonEditorConfigContributor.java:130]: Yes that's true, but we're always changing it to the same value.

But after thinking about this, I think that we should probably have a discussion about the configuration being in the Java layer and take a decision. For me this is another another argument against it.

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js (hunk 1 above)
Comment (ambrinchaudhary) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:25]: I see @julien, but my point of view here is different. Until now editors where rendered via taglib, and it was not like related to frontend people, but also backends could configure it. Current java configuration allows you easily to have different editor configurations for a custom portlet (for example in wiki you have `html` editor config and `creole` config), or differents editors in the same portlet (like blogs, where you have 3 or 4 editors at the same time, each of them different). And we can still take advantage of this current way of configuration for rendering editors via JS. If we now want to think in another way of customization for editor's configuration from the frontend point of view, we will need time to think about it, and we are adding an extra way to do that just for frontend developers (because I guess we should maintain the current way)... so... extra code to maintain. 

In conclusion, in the teams I think we are more than used to this configuration way, and... I'm not sure this is something related exclusively to frontend infrastructure team.... that is only my personal point of view, I may be wrong of course... 

So let's see if someone else have more things to add...  /cc @liferay-frontend

Comment (julien) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:25]: > Current java configuration allows you easily to have different editor configurations for a custom portlet 

You're right, but my take on it is that the Java configuration is just "copying" what CKEditor provides, at the end it's just an object with options and you can do exactly the same thing with JS.

So I agree with you that we still need to be able to configure the editor via Java, but we also need to support configuring an editor only with JS.

> and we are adding an extra way to do that just for frontend developers

That would make sense for me, since we're frontend infrastructure and not backend infrastructure.Sorry to repeat myself, but the way I see it is that we added an *extra* way to be able to use them via Java, remember: CKEditor is a JS library not a Java library.

> In conclusion, in the teams I think we are more than used to this configuration way, and

That's what you think, and it's a valid opinion, but I doubt every frontend developer thinks the same (a with me you already have an example :joy: ), so why not support both ways ?

Comment (julien) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:25]: Additionally, one thing that I don't understand is [this](https://github.com/liferay-frontend/liferay-portal/pull/960/files#diff-91a08522aea78564ab2061bea292615577dc05c3db94c579dd042dd8c5107544R62-R63), and to save a click I'll put it here

```java
String balloonToolbarPlugins =
	"balloontoolbar,floatingspace,stylescombo";

String extraPlugins = jsonObject.getString("extraPlugins");

if (Validator.isNotNull(extraPlugins)) {
	extraPlugins += "," + balloonToolbarPlugins;
}
else {
	extraPlugins = balloonToolbarPlugins;
}
```

If I understand correctly, `EditorConfigContributor`'s should 

> Provides an interface for setting the editor's configuration

(*Found [here](https://docs.liferay.com/portal/7.0/javadocs/portal-kernel/com/liferay/portal/kernel/editor/configuration/EditorConfigContributor.html), which might be "old" but it seems like a valid definition*)

And in this case, what puzzles me is that without the `balloontoolbar` (and `floatingspace`) plugins we're not going to have a "balloon editor" but a "classic editor", so I think that those two plugins should at least be in the JS (or at least always present in the editor's configuration), to make sure that no matter what "options" are provided by the `EditorConfigContributor` class we always get a "balloon editor" ... and a balloon editor without toolbars doesn't do much, so I also think we need a basic set of toolbars defined somewhere too.

I know this is a long conversation we're having here, but I'm just trying to make sure we don't miss anything.

Comment (ambrinchaudhary) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:25]: As discussed in our meeting with Daniel Sanz, there will be some basic config in the Balloon Editor that could be completed or overrided (in the toolbar cases) with the ConfigContributors. See https://github.com/liferay-frontend/liferay-portal/pull/984/commits/71bee6799b2b07e146d73249482957456b2a6a8a

**Comment by ambrinchaudhary** [comment]
Closing in favor of https://github.com/liferay-frontend/liferay-portal/pull/984
